native/src/main/java/io/ballerina/stdlib/serdes/Deserializer.java

```diff
@@ -52,251 +66,335 @@
-            return getBallerinaUnionTypeValueFromMessage(dynamicMessageForUnion, type, schema);
-        } else if (type.getTag() == TypeTags.ARRAY_TAG) {
-            ArrayType arrayType = (ArrayType) type;
-            Type elementType = arrayType.getElementType();
-            FieldDescriptor fieldDescriptor = schema.findFieldByName(ARRAY_FIELD_NAME);
-            schema = fieldDescriptor.getContainingType();
-            return getBallerinaArrayValueFromMessage(dynamicMessage.getField(fieldDescriptor), elementType, schema, 1);
-        } else if (type.getTag() == TypeTags.RECORD_TYPE_TAG) {
-            Map<String, Object> mapObject = getBallerinaRecordValueFromMessage(dynamicMessage, type, schema);
-            return ValueCreator.createRecordValue(type.getPackage(), type.getName(), mapObject);
-        } else {
-            throw createSerdesError(UNSUPPORTED_DATA_TYPE + type.getName(), SERDES_ERROR);
+            case TypeTags.UNION_TAG: {
+                return getUnionTypeValueFromMessage(dynamicMessage, ballerinaType);
+            }
+
+            case TypeTags.ARRAY_TAG: {
+                Type elementType = ((ArrayType) ballerinaType).getElementType();
+                Descriptor messageDescriptor = dynamicMessage.getDescriptorForType();
+                FieldDescriptor fieldDescriptor = messageDescriptor.findFieldByName(ARRAY_FIELD_NAME);
+                Object messageFieldValue = dynamicMessage.getField(fieldDescriptor);
+                return getArrayTypeValueFromMessage(messageFieldValue, elementType, messageDescriptor);
+            }
+
+            case TypeTags.RECORD_TYPE_TAG: {
+                return getRecordTypeValueFromMessage(dynamicMessage, (RecordType) ballerinaType);
+            }
+
+            default:
+                throw createSerdesError(UNSUPPORTED_DATA_TYPE + ballerinaType.getName(), SERDES_ERROR);
         }
     }
 
-    private static Object getBallerinaPrimitiveValueFromMessage(Object value) {
+    private static Object getDecimalPrimitiveTypeValueFromMessage(DynamicMessage decimalMessage) {
+
+        Descriptor decimalSchema = decimalMessage.getDescriptorForType();
+
+        FieldDescriptor valueField = decimalSchema.findFieldByName(VALUE);
+        FieldDescriptor scaleField = decimalSchema.findFieldByName(SCALE);
+        FieldDescriptor precisionField = decimalSchema.findFieldByName(PRECISION);
+
+        BigInteger value = new BigInteger(((ByteString) decimalMessage.getField(valueField)).toByteArray());
+        Integer scale = (Integer) decimalMessage.getField(scaleField);
+        MathContext precision = new MathContext((int) decimalMessage.getField(precisionField));
+
+        BigDecimal bigDecimal = new BigDecimal(value, scale, precision);
+        return ValueCreator.createDecimalValue(bigDecimal);
+    }
+
+    private static Object getPrimitiveTypeValueFromMessage(Object value) {
+
```

Comment:
Let's not add an empty line as the first line in a function body.
```suggestion
```

Reply:
Updated accordingly